### PR TITLE
feat: add support for fossil-scm in prompt

### PR DESCRIPTION
**Pull request by nat-418**
## Description

Add fish prompt function and include it in the vcs prompt function.

Fixes issue #9497

## TODOs:
- [ ] Changes to fish usage are reflected in user documentation/manpages.
- [ ] Tests have been added for regressions fixed
- [ ] User-visible changes noted in CHANGELOG.rst

### CODE REVIEW

File: share/functions/fish_fossil_prompt.fish [hunk 1]
```diff
@@ -0,0 +1,53 @@
+function fish_fossil_prompt --description 'Write out the fossil prompt'
+    # Bail if fossil is not available
+    if not command -sq fossil
+        return 1
+    end
+
+    # Bail if not a fossil checkout
+    if not test -e .fslckout
```
Comment (faho) [inline share/functions/fish_fossil_prompt.fish:8]: This is too strict - like git and hg looking for a .git or .hg directory, fossil looks in parent directories as well - see https://fossil-scm.org/home/doc/tip/www/tech_overview.wiki:

> For Fossil commands that run from within a working checkout, the first thing that happens is that Fossil locates the checkout database. Fossil first looks in the current directory. If not found there, it looks in the parent directory. If not found there, the parent of the parent. And so forth until either the checkout database is found or the search reaches the root of the file system. (In the latter case, Fossil returns an error, of course.) Once the checkout database is located, it is used to locate the repository database. 

So you need to basically do what `fish_print_hg_root` does, except replacing `.hg/dirstate` with `.fslckout` (and *possibly* `_FOSSIL_` - no idea when that name is used).

Alternatively, if `fossil info` is quick enough outside of a repo, say ~5ms (like git), you can remove this check and do

```fish
set -l fossil_info (fossil info)
or return
```

below (possibly with `fossil info 2>/dev/null` if it writes an error).

File: share/functions/fish_fossil_prompt.fish [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+    echo -n $fossil_project_name
+    set_color normal
+    echo -n ':'
+    set_color --bold yellow
+    echo -n $fossil_tags
+    set_color normal
+
+    # Parse fossil status
+    set -l fossil_status (fossil status)
+    if string match --index --regex --quiet '^ADDED' $fossil_status
```
Comment (faho) [inline share/functions/fish_fossil_prompt.fish:30]: I don't know what the `--index` is for, it seems to not do anything.

Comment (faho) [reply share/functions/fish_fossil_prompt.fish:30]: You can also use glob-mode here:

```fish
if string match --quiet 'ADDED*' $fossil_status
```

which is a bit shorter.

**Review by faho — CHANGES_REQUESTED** [review]
1. The in-repository will fail in subdirectories
2. Some style nits

**Comment by faho** [comment]
As a general question: How long does this take inside of a repo with fossil installed?

File: share/functions/fish_fossil_prompt.fish (hunk 1 above)
Comment (faho) [reply share/functions/fish_fossil_prompt.fish:8]: From what I can find `_FOSSIL_` was the old name, and it's been renamed before 2016. So it's fine to ignore for our use, if we need to check for the checkout database ourselves.

**Comment by nat-418** [comment]
> As a general question: How long does this take inside of a repo with fossil installed?
> 
> You can check with `time fish_fossil_prompt`.
```
$ time fish_fossil_prompt
Executed in   84.28 millis    fish           external
   usr time   13.72 millis    3.16 millis   10.56 millis
   sys time    9.85 millis    0.01 millis    9.84 millis

```

**Comment by nat-418** [comment]
> 1. The in-repository will fail in subdirectories
> 
>     2. Some style nits

Thanks for the quick and helpful feedback. I apologize for forgetting to mark this PR as a draft when I opened it.

File: share/functions/fish_fossil_prompt.fish (hunk 2 above)
Comment (nat-418) [reply share/functions/fish_fossil_prompt.fish]: changed

File: share/functions/fish_fossil_prompt.fish (hunk 1 above)
Comment (nat-418) [reply share/functions/fish_fossil_prompt.fish]: changed

**Comment by faho** [comment]
Merged, thanks!

For the record `fossil ls` outside of a fossil repo takes ~3-4ms, so it's about as fast as `git`. We *could* use builtins here like hg, but it's a much different trade-off (`hg status` is 150ms outside of a repo, *hot cache*).

**Comment by nat-418** [comment]
Thank you! I love fish!

**Pull request merged**